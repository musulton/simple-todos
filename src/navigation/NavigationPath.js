export const SPLASH_PATH = 'SPLASH';
export const LOGIN_PATH = 'LOGIN';
export const TODO_PATH = 'TODO';
