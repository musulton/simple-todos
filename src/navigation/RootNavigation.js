import {createNavigationContainerRef, StackActions} from '@react-navigation/native';

export const navigationRef = createNavigationContainerRef()

export function navigate(name, params, isReplace = false) {
    if (navigationRef.isReady()) {
        if (isReplace) {
            navigationRef.current.dispatch(StackActions.replace(name))
        } else {
            navigationRef.navigate(name, params);
        }

    }
}
