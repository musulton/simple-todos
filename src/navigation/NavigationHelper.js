import * as RootNavigation from './RootNavigation';

export function goToScreen(routePath, isReplace) {
    RootNavigation.navigate(routePath, null, isReplace);
}
