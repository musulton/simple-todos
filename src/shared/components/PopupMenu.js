import {Modal, Text, StyleSheet, View, TouchableOpacity, TouchableWithoutFeedback, Pressable} from "react-native";
import React, {useEffect, useState} from "react";
import {useDispatch, useSelector} from "react-redux";
import Icon from "@expo/vector-icons/FontAwesome";

import {LOGIN_PATH} from "../../navigation/NavigationPath";
import {logout} from "../store/login/LoginAction";

const PopupMenu = ({navigation}) => {
    const dispatch = useDispatch();
    const isLoggedIn = useSelector((state) => state.LoginReducer.isLoggedIn);
    const [modalVisible, setModalVisible] = useState(false);
    useEffect(() => {
        if (isLoggedIn === false) {
            navigation.replace(LOGIN_PATH);
        }
    })
    return (
        <View>
            <Pressable style={{paddingLeft: 30}}
                       onPress={() => setModalVisible(!modalVisible)}
            >
                <Icon name={'ellipsis-v'}
                      size={16}/>
            </Pressable>
            <Modal
                animationType="fade"
                transparent={true}
                visible={modalVisible}
                onRequestClose={() => {
                    setModalVisible(!modalVisible);
                }}
            >
                <TouchableOpacity style={{flex: 1}} activeOpacity={1}
                                  onPressOut={() => {
                                      setModalVisible(false)
                                  }}>
                    <View style={styles.centeredView}>
                        <TouchableWithoutFeedback>
                            <View style={styles.modalView}>
                                <Pressable onPress={() => {
                                    dispatch(logout());
                                }}>
                                    <Text style={styles.modalText}>Logout</Text>
                                </Pressable>
                            </View>
                        </TouchableWithoutFeedback>
                    </View>
                </TouchableOpacity>
            </Modal>
        </View>
    )
}

const styles = StyleSheet.create({
    centeredView: {
        flex: 1,
        justifyContent: "flex-start",
        alignItems: "flex-end",
        marginTop: 22
    },
    modalView: {
        margin: 20,
        backgroundColor: "white",
        borderRadius: 10,
        padding: 32,
        alignItems: "center",
        shadowColor: "#000",
        shadowOffset: {
            width: 0,
            height: 2
        },
        shadowOpacity: 0.25,
        shadowRadius: 4,
        elevation: 5
    },
    modalText: {
        textAlign: "center"
    }
});


export default PopupMenu;
