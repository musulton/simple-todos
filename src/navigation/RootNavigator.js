import React from "react";
import {NavigationContainer} from "@react-navigation/native";
import {createNativeStackNavigator} from "@react-navigation/native-stack";

import LoginScreen from "../screens/LoginScreen/LoginScreen"
import ToDoScreen from "../screens/TodoScreen/ToDoScreen";
import SplashScreen from "../screens/SplashScreen/SplashScreen";
import {LOGIN_PATH, SPLASH_PATH, TODO_PATH} from "./NavigationPath";
import PopupMenu from "../shared/components/PopupMenu";
import LoginService from "../services/LoginService";
import {Login} from "../screens/LoginScreen/Login";
import {navigationRef} from "./RootNavigation";

const Stack = createNativeStackNavigator();

const RootNavigator = () => {
    return <NavigationContainer ref={navigationRef}>
        <Stack.Navigator initialRouteName={SPLASH_PATH}>
            <Stack.Screen name={SPLASH_PATH} component={SplashScreen} options={{headerShown: false}}/>
            <Stack.Screen name={LOGIN_PATH} options={{headerShown: false}}>
                {props => <LoginScreen {...props} login={() => Login(LoginService)} />}
            </Stack.Screen>
            <Stack.Group
                screenOptions={({navigation}) => ({
                    headerStyle: {
                        backgroundColor: '#f2f2f2',
                    },
                    headerShadowVisible: false,
                    headerRight: () => <PopupMenu navigation={navigation}/>
                })}
            >
                <Stack.Screen name={TODO_PATH} component={ToDoScreen} options={{ title: '' }}/>
            </Stack.Group>
        </Stack.Navigator>
    </NavigationContainer>
}

export default RootNavigator;
