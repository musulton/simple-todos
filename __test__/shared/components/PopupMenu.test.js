import {render} from "@testing-library/react-native";

import PopupMenu from "../../../src/shared/components/PopupMenu";
import {useSelector} from "react-redux";
import {LOGIN_PATH} from "../../../src/navigation/NavigationPath";

jest.mock("react-redux", () => ({
    useSelector: jest.fn(),
    useDispatch: jest.fn()
}))

describe("PopupMenu", () => {
    test("should navigate to LOGIN_PATH when isLoggedIn is false", () => {
        useSelector.mockImplementation(() => false)
        const mockProps = {
            navigation: {
                replace: jest.fn()
            }
        }

        render(<PopupMenu {...mockProps} />)

        expect(mockProps.navigation.replace).toHaveBeenCalledWith(LOGIN_PATH)
    })
})
