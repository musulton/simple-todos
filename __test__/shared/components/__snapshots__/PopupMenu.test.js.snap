// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupMenu should render correctly 1`] = `
<View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "paddingLeft": 30,
      }
    }
  >
    <Text />
  </View>
  <Modal
    animationType="fade"
    data-testid="modal"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    transparent={true}
    visible={false}
  />
</View>
`;
